src/journal/journal-fields.c

```diff
@@ -0,0 +1,363 @@
+        siphash24_compress(((StringKey*)p)->sz, ((StringKey*)p)->len, state);
+};
+
+static int string_key_compare_func(const void *a, const void *b) {
+        int r = strncmp(((StringKey*)a)->sz, ((StringKey*)b)->sz, MIN(((StringKey*)a)->len, ((StringKey*)b)->len));
+        if (r==0) { return ((StringKey*)b)->len - ((StringKey*)a)->len; }
+        return r;
+};
+
+const struct hash_ops string_key_hash_ops = {
+        .hash = string_key_hash_func,
+        .compare = string_key_compare_func
+};
+
+/* ============================================================ */
+
+/* A hash-table entry, which contains another hash-table
+   and some counters
+*/
+
+typedef struct _HashN {
+        size_t n;   /* number of entries having this field      */
+        size_t nv;  /* number of distinct values for this field */
+        Hashmap *h; /* hash of distinct values for this field   */
+} HashN;
+
+/* ============================================================ */
+
+/* Put a field value (sk) into the hash of values for this field */
+
+static int put_journal_value(HashN *hh, StringKey sk) {
+        int r = 0;
+        size_t nn;
+        StringKey *psk;
+
+        /* Try to find the field value in the hash */
+        nn = PTR_TO_SIZE(hashmap_get(hh->h, &sk));
+
+        if (nn == 0) {
+                /* The field value was NOT found, so add it */
+
+                hh->nv++;
+
+                psk = malloc(sizeof(StringKey));
+                if (!psk) { return -ENOMEM; }
+                psk->len = sk.len;
+                psk->sz = malloc(sk.len+1);
+                if (!(psk->sz)) { return -ENOMEM; }
+                memcpy(psk->sz, sk.sz, sk.len);
+                psk->sz[sk.len] = '\0';
+
+                r = hashmap_put(hh->h, psk, SIZE_TO_PTR(1));
+                if (r < 0) { return r; }
+
+        } else {
+                /* The field value WAS found, so update the counter */
+
+                r = hashmap_update(hh->h, &sk, SIZE_TO_PTR(nn+1));
+                if (r < 0) { return r; }
+        }
+        return 0;
+}
+
+/* ============================================================ */
+
+/* Handle journalctl --fields */
+
+int journal_fields(sd_journal *j, int n, const char* field) {
+        int r = 0;
+        size_t n_entries=0;
+        Hashmap *h = NULL;
+        HashN *hh;
+        const void *data;
+        size_t length;
+        const char *eq;
+        Iterator i;
+        Iterator i2;
+        size_t nn;
+        unsigned u;
+        unsigned n_columns;
+        StringKey sk;
+        StringKey *psk;
+        StringKey *psk2;
+        size_t max_field_len = 1;
+        int num_field_width;
+        const char *csz;
+        char *sz;
+        char *sz1;
+        char *sz2;
+        const char *color_on = "", *color_off = "";
+        double pct;
+
+        r = sd_journal_set_data_threshold(j, 0);
+        if (r < 0) { return log_error_errno(r, "Failed to unset data size threshold: %m"); }
```

Comment:
I can't speak to the likelihood of this landing, but at a glance the code here isn't compliant with the systemd coding style.  Read https://github.com/systemd/systemd/blob/master/CODING_STYLE and at least try be compliant before posting a PR, it will save everyone involved a round trip.